File: skills.py
from flask import Flask, jsonify, request
from flask_cors import CORS
import mysql.connector

app = Flask(__name__)
CORS(app)

# Database Connection
def get_db_connection():
    return mysql.connector.connect(
        host="localhost",
        user="root",
        password="Suhani@2804",
        database="minortable1"
    )

# Fetch Courses Based on Skill & Platform
@app.route("/api/get_courses", methods=["GET"])
def get_courses():
    skill = request.args.get("skill")
    platform = request.args.get("platform")

    if not skill or not platform:
        return jsonify({"error": "Skill and Platform are required"}), 400

    conn = get_db_connection()
    cursor = conn.cursor(dictionary=True)
    
    # Modify the SQL query to use LIKE for partial matching
    cursor.execute("SELECT title, course_link FROM Courses WHERE search_query LIKE %s AND platform = %s", (f"%{skill}%", platform))
    
    courses = cursor.fetchall()
    conn.close()

    # Ensure every course has a proper link, fallback to platform search URL if missing
    for course in courses:
        if not course["course_link"]:
            course["course_link"] = get_search_url(platform, skill)

    return jsonify({"courses": courses})

def get_search_url(platform, skill):
    search_links = {
        "Coursera": f"https://www.coursera.org/search?query={skill}",
        "Udemy": f"https://www.udemy.com/courses/search/?q={skill}",
        "Unstop": f"https://unstop.com/courses?search={skill}",
        "edX": f"https://www.edx.org/search?q={skill}",
        "FutureLearn": f"https://www.futurelearn.com/search?q={skill}",
        "Skillshare": f"https://www.skillshare.com/search?query={skill}",
        "LinkedIn Learning": f"https://www.linkedin.com/learning/search?keywords={skill}",
        "Pluralsight": f"https://www.pluralsight.com/search?q={skill}",
        "Khan Academy": f"https://www.khanacademy.org/search?page_search_query={skill}"
    }
    return search_links.get(platform, "https://google.com")

if __name__ == "__main__":
    app.run(port=5000, debug=True)

File: skills.js
document.addEventListener("DOMContentLoaded", () => {
    const skillDropdown = document.getElementById("skill-dropdown");
    const user = JSON.parse(localStorage.getItem("user"));

    skillDropdown.innerHTML = "";

    const allOption = document.createElement("option");
    allOption.value = "all";
    allOption.textContent = "All Skills";
    skillDropdown.appendChild(allOption);

    if (user && user.skills) {
        const skillsArray = user.skills.split(',').map(skill => skill.trim());
        skillsArray.forEach(skill => {
            const option = document.createElement("option");
            option.value = skill;
            option.textContent = skill;
            skillDropdown.appendChild(option);
        });
    } else {
        skillDropdown.innerHTML += "<option value=''>No skills found</option>";
    }
});

function fetchCourses() {
    const skill = document.getElementById("skill-dropdown").value;
    const platform = document.getElementById("platform").value;
    const courseContainer = document.getElementById("course-list");

    if (!skill) {
        alert("Please select a skill.");
        return;
    }

    document.getElementById("loading").style.display = "block";
    courseContainer.innerHTML = "";

    fetch(`http://127.0.0.1:5000/api/get_courses?skill=${encodeURIComponent(skill)}&platform=${encodeURIComponent(platform)}`)
        .then(response => response.json())
        .then(data => {
            document.getElementById("loading").style.display = "none";
            courseContainer.innerHTML = "";

            if (!data.courses || data.courses.length === 0) {
                let noCourseMsg = document.createElement("p");
                noCourseMsg.textContent = "No courses found. Click below to explore more.";
                courseContainer.appendChild(noCourseMsg);
            } else {
                let courseTitles = new Set();

                data.courses.forEach(course => {
                    if (!courseTitles.has(course.title)) {
                        courseTitles.add(course.title);

                        const card = document.createElement("div");
                        card.className = "course-card";

                        const courseLink = course.course_link || getSearchUrl(platform, skill);

                        card.innerHTML = `
                            <h3>${course.title}</h3>
                            <p><strong>Platform:</strong> ${platform}</p>
                            <a href="${courseLink}" target="_blank">View Course</a>
                        `;

                        courseContainer.appendChild(card);
                    }
                });
            }

            const moreCoursesBtn = document.createElement("button");
            moreCoursesBtn.textContent = "Look for More Courses";
            moreCoursesBtn.classList.add("more-courses-btn");
            moreCoursesBtn.onclick = () => {
                window.open(getSearchUrl(platform, skill), "_blank");
            };
            courseContainer.appendChild(moreCoursesBtn);
        })
        .catch(error => {
            console.error("Error fetching courses:", error);
            document.getElementById("loading").style.display = "none";
            courseContainer.innerHTML = "<p>Error fetching courses. Please try again later.</p>";
        });
}

function getSearchUrl(platform, skill) {
    if (platform === "all") {
        return `https://www.google.com/search?q=${encodeURIComponent(skill + ' online courses')}`;
    }

    const searchLinks = {
        "Coursera": `https://www.coursera.org/search?query=${encodeURIComponent(skill)}`,
        "Udemy": `https://www.udemy.com/courses/search/?q=${encodeURIComponent(skill)}`,
        "Unstop": `https://unstop.com/courses?search=${encodeURIComponent(skill)}`,
        "edX": `https://www.edx.org/search?q=${encodeURIComponent(skill)}`,
        "FutureLearn": `https://www.futurelearn.com/search?q=${encodeURIComponent(skill)}`,
        "Skillshare": `https://www.skillshare.com/search?query=${encodeURIComponent(skill)}`,
        "LinkedIn Learning": `https://www.linkedin.com/learning/search?keywords=${encodeURIComponent(skill)}`,
        "Pluralsight": `https://www.pluralsight.com/search?q=${encodeURIComponent(skill)}`,
        "Khan Academy": `https://www.khanacademy.org/search?page_search_query=${encodeURIComponent(skill)}`
    };

    return searchLinks[platform] || "https://www.google.com";
}
